### ISSUE TITLE
Add rubocop style progress reporting to reek

### ISSUE THREAD

**Issue by jhubert** [troessner/reek#980]
I'm a big fan of the way that RuboCop reports on the progress of an inspection:

![image](https://cloud.githubusercontent.com/assets/3384/16897438/8a25bb14-4b66-11e6-950f-f6e0b156e99b.png)

It's especially great when using RuboCop with something like Guard that runs every time a file is changed.

That style of progress display would be really nice in Reek.

**Comment by troessner** [comment]
>> I'm a big fan of the way that RuboCop reports on the progress of an inspection:

Great initiative and I fully agree :)

I did check out both of your pull requests but I think it makes sense if we discuss the specifics of this feature here before going into code since the same questions pop up and this will be pretty hard to follow and discuss on 2 different pull requests.

**The following questions and discussion items come to my mind:**

* Should the "progress bar" be a property of the "text format" or it's own format? You modelled it [here](https://github.com/troessner/reek/pull/983/files#diff-ccf18235057177600222fe6eedf2f132R111) as its own format (albeit directly inheriting from TextReport) which I personally find a little confusing. I'd rather make it just a property of TextReport which means it would be a separate flag like [in your other pull request](https://github.com/troessner/reek/pull/982/files#diff-b5b9b45d56bae1cd97fb9ee6745139d4R49).
* Looking at [making it part of the ReportCommand](https://github.com/troessner/reek/pull/982/files#diff-9ce866437f22b59ed4a1995bf2c65aadR23) I believe the right way to go is indeed to put this behaviour into our reports
* Given that the "progress bar" is only available for text reports, wouldn't it make sense to enable it by default?

cc @mvz @waldyr

**Comment by mvz** [comment]
Looking at it from a functional level, I would have it appear as an option for the text report, unless we have it appear on STDERR. In the latter case, we can have it everywhere and it won't interfere with the report output.

I see that RuboCop has it as a separate report format, but that is indeed a bit odd.

If we go for the option of putting the behavior in TextReport, we can always model it as a decorator or put it in the superclass if we want to make it available for more than one format.

It makes sense to enable the progress reporter by default eventually, but I think that should not be done in the 4.x series.

TL;DR: My vote is to make it an non-default option on TextReport and implement it there.

**Comment by troessner** [comment]
>> It makes sense to enable the progress reporter by default eventually, but I think that should not be done in the 4.x series.

Why? I really don't think we need to be so anal about semver for Reek. This feature is a clear improvement and I think we should ship as fast as possible.

This is not a breaking change - why not let the users benefit from things like that as fast as possible?

**Comment by mvz** [comment]
> Why?

We did get a fair number of bug reports the last time we changed the text output. I'm thinking e.g., the vim plugin could have problems.

If we output the progress to STDERR this would be less of a problem, but I don't know how you feel about that.

**Comment by troessner** [comment]
>> We did get a fair number of bug reports the last time we changed the text output. I'm thinking e.g., the vim plugin could have problems.

But that's more kind of a bug than really a breaking change, is it? If we do it like that it's like saying "we're declaring this a breaking change because it **probably** will trigger some bugs", isn't it?

I'm really wary of not releasing features for reasons like that. We should ship as fast as possible, cause that's part of the fun in non-commercial software, isn't it?
We usually fix critical bugs **really** fast so I'd rather be progressive than conservative.

How about just increasing our (manual) test coverage for this feature? 
We can easily check all the associated plugins (there aren't that many) how they deal with this change and then release if we have a good feeling about it.

**Comment by mvz** [comment]
I meant bug reports from people relying on the old format. But yeah, :+1: on just checking all the plugins and then making it the default.

**Comment by troessner** [comment]
On second thought and looking at [all our editor integrations](https://github.com/troessner/reek#editor-integrations) I'd like to change my opinion and follow more @mvz approach, meaning:

```
If we go for the option of putting the behavior in TextReport, we can always model it as a decorator or put it in the superclass if we want to make it available for more than one format.
```

And let's make it default right away.

We should then release 5.0.0 right away to get this feature out on the road asap (so not wait until we accumulated other breaking features because that's something I'd consider toxic for developer fun)

Everybody on board with this?

**Comment by mvz** [comment]
Sounds good to me.

**Comment by troessner** [comment]
@jhubert I'm sorry that this feature got dragged out by us apparently. Can I help you to get this going?
If you like we can pair next week for an hour to get this ready for merge.

**Comment by jhubert** [comment]
Ugh. Sorry for holding this up guys. We're rolling out to our largest customer ever at work and it's been crazy.

@troessner I'll take a crack at it tonight. My impression is that it's just the tests that need fixing?

**Comment by troessner** [comment]
>> @troessner I'll take a crack at it tonight. My impression is that it's just the tests that need fixing?

Yes, just the specs! There are a couple of minor things that have changed in the mean time (e.g. we refactored our sample files so that we don't have to do inline code samples anymore and just can [use our sample steps](https://github.com/troessner/reek/blob/master/features/step_definitions/sample_file_steps.rb#L3)) but that is something I can update after we merged this one.
So let's rather focus on getting this merged to not drag this out even longer for you.
Please also squash to one commit after you fixed the specs and then I'm happy to merge this one right away :)

**Comment by backus** [comment]
Hey guys I just wanted to chime in and say I love this new reporter. I think it is a great addition and I appreciate the work that went into it.

Also this issue can probably be closed now, right?

**Comment by troessner** [comment]
It can!
